Add DPI scaling factor and revert content scaling factor behaviour

vadz opened the pull request:
This reverts `wxWindow::GetContentScaleFactor()` back to what it did in 3.0 and adds new `GetDPIScalingFactor()` which can be used to retrieve the DPI scaling on all platforms.

This allows to get rid of some preprocessor checks (although not all, unfortunately).

vadz committed:
Add wxWindow::GetDPIScaleFactor()

This function replaces some uses of GetContentScaleFactor(), where a
factor greater than 1 must be used even under the platforms not doing
any logical/physical pixel mapping, such as MSW.

For now GetContentScaleFactor() is still unchanged, but it will return 1
for such platforms in the future and adding GetDPIScaleFactor() allows
to avoid changing the behaviour of the code which relied on its current
behaviour.

vadz committed:
Make wxWindow::GetContentScaleFactor() return 1 under MSW again

This reverts bc492a9e6e (Make wxWindow::GetContentScaleFactor() useful
for non-OSX platforms., 2015-03-18) and restores the old behaviour from
wxWidgets 3.0, which consisted in only returning factor different from 1
from this function for the platforms distinguishing logical and physical
pixels.

After this change, the return value of this function can be portably
used on all platforms to convert between logical and physical pixels,
independently of the current DPI.

vadz committed:
Remove recently added GetOpenGLScaleFactor()

It has become unnecessary after the previous commit, as now the generic
GetContentScaleFactor() can be used instead of it on all platforms, so
revert the changes of f6cc8ff52c (Add GetOpenGLScaleFactor() to abstract
OpenGL coordinates scaling, 2020-07-10).

See https://github.com/wxWidgets/wxWidgets/pull/1944

See #17391.

vadz commented:
@MaartenBent As I wrote in wx-dev (but this thread is long and I'm afraid you might miss it there, so I'm duplicating it here), this PR makes behaviour of `GetContentScaleFactor()` in `wxWindow` and `wxDC` inconsistent under MSW, because the former now always returns 1, while the latter can return something bigger. I believe that b0152155c0 (Implement GetContentScaleFactor for wxMSWDCImpl, 2019-08-02) should be reverted too as part of this, but I'm not totally sure about this. Still, if I look at e.g. `wxSharedDCBufferManager::DoCreateBuffer()`, it seems to me that it should really return 1 under MSW because we should always create a bitmap of the same size as the window there, shouldn't we?

csomor commented:
I'm fine with these changes, actually I like having things under one model again :-) Thanks

do you want to integrate my GetDPI fix for macOS/iOS from my PR https://github.com/wxWidgets/wxWidgets/pull/1984 here, or should I just add things after this has been merged

A few thoughts and comments

- the 'platform default' or baseline on msw is 96 dpi but on mac it is 72 dpi, perhaps we should mention this somewhere as right now 96 (use constants not numbers :-) is mentioned in a few places ...

- could we have some pixels and high dpi overview chapter explaining chapter there we could explain in one place the context and how things are related, and from the API we could refer to this, having all 3 explained
  - Device-Independent pixels - (DIP, very dangerously close to DPI) - using a 'virtual' baseline grid coordinate system of platform default dpi over all the existing screens
  - Logical pixels - (LOGP ?) the coordinate system used for drawing to DCs, sizing controls etc
  - Physical pixels (PHYP ?) the truly physical pixels coordinate system, hardly used for the 'average' wx dev

- I really, really, think that having a GetDIPScaleFactor is important for keeping conversion code simple

  `GetDIPScaleFactor() { return GetDPScaleFactorI() / ContentScaleFactor(); }`

- from my point of view GetContentScale on the dc should be the same as on window, but I haven't worked with the msw details in high-dpi-land so I may be overlooking something

MaartenBent commented:
I added b015215 to fix the font size in [wxSVGFileDC](https://github.com/wxWidgets/wxWidgets/blob/master/src/common/dcsvg.cpp#L361) on a high DPI screen in wxMSW.
 There the PPI of the font is based on `wxScreenDC`, but since `wxSVGFileDC` should be DPI independent, we want it at 96 (on Windows at least).
It can be replaced with `const double scale = sDC.GetPPI().y / 96.0;`, and wrap the scaling with `#ifdef __WXMSW__`.

csomor commented:
> I added [b015215](https://github.com/wxWidgets/wxWidgets/commit/b0152155c0cd48d38004fc246a2540cd5fec438f) to fix the font size in [wxSVGFileDC](https://github.com/wxWidgets/wxWidgets/blob/master/src/common/dcsvg.cpp#L361) on a high DPI screen in wxMSW.
> There the PPI of the font is based on `wxScreenDC`, but since `wxSVGFileDC` should be DPI independent, we want it at 96 (on Windows at least).
> It can be replaced with `const double scale = sDC.GetPPI().y / 96.0;`, and wrap the scaling with `#ifdef __WXMSW__`.

and this worked correctly on retina displays ? what's the easiest way to test this ? I'd think that this might be a case for using my GetDIPScaleFactor instead on all platforms, but I have to test ...

vadz committed:
Revert "Implement GetContentScaleFactor for wxMSWDCImpl"

This reverts commit b0152155c0cd48d38004fc246a2540cd5fec438f.

After changing wxWindow::GetContentScaleFactor() to return 1 on
platforms without logical pixels, such as MSW, in the grandparent
commit, make wxDC::GetContentScaleFactor() consistent with it too.

MaartenBent commented:
I don't have the ability to test it on macOS with a retina display. Maybe you can compare it to this svg image (remove `.txt`, github doesnt like `.svg`) from the SVG sample, text tab.
[Text.svg.txt](https://github.com/wxWidgets/wxWidgets/files/4943637/Text.svg.txt)

vadz commented:
@csomor This PR doesn't (or at least isn't supposed to) change anything for Mac/GTK 3, so I don't think the change from #1984 really belongs here, I'd rather merge this one.

We definitely need "High DPI Support" overview page in the manual containing everything you mention as well as platform-specific things, such as manifest under MSW or `Info.plist` keys [mentioned here](https://developer.apple.com/library/archive/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/Explained/Explained.html). I won't have time to write it before 3.1.4 however :-( 

We can add `GetDIPScaleFactor()` but I'd really like to have a concrete example of when it would be useful to put it in the documentation. Right now I don't really see it. The name looks good to me, however: as it would return DPI scaling under MSW and always 1 under Mac/GTK 3, it's indeed the factor by which `FromDIP()` multiplies its argument. OTOH this also explains why I don't think it's really useful: `x*GetDIPScaleFactor()` seems to be exactly the same as `FromDIP(x)`, so are you sure we really need yet another scale factor function? I'm rather tepid about it because this would improve the probability of using a wrong function by 33% (assuming people just pick a random function when they have no idea about which one to use, which looks like a reasonable hypothesis IME).

@MaartenBent Thanks for replying! I'd like to avoid preprocessor checks, of course, but I think it's better to have one in wxSVG code than return (what seems to me like a) wrong value from `wxDC::GetContentScaleFactor()`. However I've just checked the file produced by SVG sample with and without b0152155c0cd48d38004fc246a2540cd5fec438f and somehow I don't see any difference between SVGs generated when the sample is running without or with 200% scaling: the generated SVG file (for the "Text" page) appears identical and looking at it in text editor I see that it is indeed identical, e.g. the first text line iis
```xml
  <text x="50.00" y="42.00" textLength="147" style="font-family:Arial; font-weight:400; font-style:normal; font-size:9.00pt; text-decoration: none; fill:#000000; fill-opacity:1.00; stroke:#000000; stroke-opacity:1.00; stroke-width:0; white-space: pre;" xml:space="preserve" transform="rotate(0.00 50.00 42.00)">This is a Swiss-style string</text>
```
in both cases. I don't know what's going on here, but it looks like DPI scaling is getting already compensated somewhere?

So for now I've just added the commit making `wxDC::GetContentScaleFactor()` return 1 too, but please let me know if this breaks something. BTW, I've also tested it under GTK with `GDK_SCALE=2` and the SVG also looks correct there, although not quite identical as there are small differences in text metrics, e.g. the line above comes out as
```xml
  <text x="50.00" y="43.00" textLength="173" style="font-family:Sans; font-weight:400; font-style:normal; font-size:10.00pt; text-decoration: none; fill:#000000; fill-opacity:1.00; stroke:#000000; stroke-opacity:1.00; stroke-width:0; white-space: pre;" xml:space="preserve" transform="rotate(0.00 50.00 43.00)">This is a Swiss-style string</text>
```
there.

MaartenBent commented:
> I don't see any difference between SVGs generated when the sample is running without or with 200% scaling: the generated SVG file (for the "Text" page) appears identical and looking at it in text editor I see that it is indeed identical, e.g. the first text line iis

Is your main screen, and/or the DPI when you logged in, at 100%/96DPI? That could explain it. This is the svg saved at 150% for me:
[Text2.svg.txt](https://github.com/wxWidgets/wxWidgets/files/4943651/Text2.svg.txt)

vadz commented:
Ah, yes, of course, sorry for not realizing that it's the main display DPI that matters here and not the one of the main window.

So what we really need here is to divide by content scale factor of the main screen but only when not using logical pixels, right? I.e. this code shouldn't be necessary under Mac/GTK 3 because the returned text extents there shouldn't depend on the DPI anyhow.

vadz commented:
Sorry, just made the 200% display my main one, but somehow I still don't see the problem, i.e. I still get the same SVG generated. I don't understand what am I doing wrong, but will try to debug it a bit later.

csomor commented:
> We can add `GetDIPScaleFactor()` but I'd really like to have a concrete example of when it would be useful to put it in the documentation. Right now I don't really see it. The name looks good to me, however: as it would return DPI scaling under MSW and always 1 under Mac/GTK 3, it's indeed the factor by which `FromDIP()` multiplies its argument. OTOH this also explains why I don't think it's really useful: `x*GetDIPScaleFactor()` seems to be exactly the same as `FromDIP(x)`, so are you sure we really need yet another scale factor function? 

yes, because FromDIP is a method, so it has to be called on every parameter, perhaps even going through pointers, to arrive at the right control, this way you can get it once and apply it to whatever you like, directly, coordinates, points, sizes, you can even transfer it to another construct, like a dc

csomor commented:
> @csomor This PR doesn't (or at least isn't supposed to) change anything for Mac/GTK 3, so I don't think the change from #1984 really belongs here, I'd rather merge this one.

ok, perfect, I just wait with mine

csomor commented:
> Text.svg.txt

thanks, indeed there's a bug on my side - wxScreenDC is always returning 1.0 as ContentScale even on retina - which obviously is wrong and GetDIPScale is a better choice here - that's why its not that far off (correct on the left, mac on the right)

![Bildschirmfoto 2020-07-19 um 15 07 10](https://user-images.githubusercontent.com/470914/87875519-f5f45600-c9d1-11ea-84e4-cbf5101fdeb0.png)

MaartenBent commented:
> If I now correctly scale the text dimensions in wxSVGFileDCImpl::DoDrawRotatedText by 96.0/72.0 = 1.33 then things look better
> 
> but I guess this scale would have to applied at quite a few places, just using a global transformation would not be sufficient, because the font itself is defined in points not pixels ...

The only platform dependent code in the dcsvg renderer is `wxScreenDC`, used for `DoGetTextExtent` and `GetChar[Width,Height]`. I think this should be the only place where a fix is needed. Similar to the fix I require on Windows`.

vadz committed:
Fix units produced by wxSVGFileDC after the previous commit

Reimplement the logic of 16a02e6338 (Use DPI independent text size in
wxSVGFileDC, 2019-08-06) without using GetContentScaleFactor(), but
using wxDC::GetPPI() directly and do it under the platforms not using
logical pixels only.

This makes the units correct again in SVGs produced when using high DPI
under MSW even although wxDC::GetContentScaleFactor() now returns 1 in
this case.

vadz commented:
OK, the reason I didn't see the problem was that apparently you need to log off/back again to change the DPI reported by `wxScreenDC`, just changing the primary monitor is _not_ enough (which actually makes sense knowing how MSW handles this..), sorry for the noise.

Now that I can see the problem, and hence test the fix, I've done it following the proposal to use `wxDC::GetPPI()` and, unsurprisingly, it works as expected. I've also changed the code to not do this for Mac/GTK 3 at all because, as I've tried explaining in the comment, this really shouldn't be necessary there. OTOH I believe we should use `scale = 0.75` for Mac because of the base DPI difference, but I'm leaving this to someone else to do later, after this PR is merged.

Please let me know if there are any objections, otherwise I'll do it tomorrow.
